torch/ao/quantization/experimental/linear.py

```diff
@@ -0,0 +1,180 @@
+    to support APoT quantization.
+    We adopt the same interface as `torch.nn.Linear`, see
+    https://pytorch.org/docs/stable/nn.html#torch.nn.Linear for documentation.
+
+    Similar to :class:`~torch.nn.Linear`, attributes will be randomly
+    initialized at module creation time and will be overwritten later
+
+    Attributes:
+        alpha: `alpha` qparam of output Quantized Tensor, type: Tensor
+        gamma: `gamma` qparam of output Quantized Tensor, type: Tensor
+        quantization_levels: `quantization_levels` qparam of output Quantized Tensor, type: Tensor
+        level_indices: `level_indices` qparam of output Quantized Tensor, type: Tensor
+    """
+
+    def __init__(self, weight2quantize: torch.Tensor, signed=False):
+        # self.in_features = in_features
+        # self.out_features = out_features
+
+        super().__init__()
+
+        # hard code to APoT paper example inputs: b=4, k=2
+        observer = APoTObserver(b=4, k=2)
+
+        observer(weight2quantize)
+
+        self.alpha, self.gamma, self.quantization_levels, self.level_indices = observer.calculate_qparams(signed=signed)
+
+        quantized = quantize_APoT(weight2quantize, self.alpha, self.gamma, self.quantization_levels, self.level_indices)
+        self.weight = dequantize_APoT(quantized)
+        self.weight = self.weight.reshape(weight2quantize.shape)
+
+    def decompose_APoT(self):
+        r"""
+        Helper function to decompose sums of PoT from APoT quantization
+        into corresponding tuples of PoT terms.
+        """
+        size = int(math.sqrt(torch.numel(self.weight)))
+
+        # decompose APoT weight
+        p_all = []
+
+        # create the "levels"
+        p0 = torch.tensor([0, 2**0, 2**-2, 2**-4])
+        p1 = torch.tensor([0, 2**-1, 2**-3, 2**-5])
+
+        # make pair of tensors that associate tuple (decomp1, decomp2) with value
+        all_levels_decomposed = []
+        vals = []
+        for level0 in range(size):
+            for level1 in range(size):
+                all_levels_decomposed.append((p0[level0], p1[level1]))
+                vals.append(self.gamma * (p0[level0] + p1[level1]))
+
+        # decompose levels in weight
+        levels_decomposed = []
+        for ele in self.weight.flatten():
+            idx = vals.index(ele)
+            levels_decomposed.append(all_levels_decomposed[idx])
+
+        # reshape decomposed levels array into 2d matrix size
+        np_levels_decomposed = np.empty(len(levels_decomposed), dtype=object)
+        np_levels_decomposed[:] = levels_decomposed
+        np_levels_decomposed = np_levels_decomposed.reshape(size, size)
+
+        return np_levels_decomposed
+
+    def linear_APoT_fn(self, activation: torch.Tensor) -> torch.Tensor:
+        r"""
+        Multiply APoT quantized weight and uniformly quantized activation
+        with bitshifting instead of matrix multiplication.
+        Args:
+            activation (Tensor): uniformly quantized activation tensor
```

Comment:
can you specify the dtype of the activation, a uniformly quantized activation tensor would have dtype quint8, whereas the test takes in a uint8, it should also be noted that simply bit shifting the integer in a quint8 will not give the correct output when z!=0 but if youo also bitshift z then it should work, though the recombination will be problematic.

In general i'm not sure if this makes sense to combine with uniform quantization since its not a fixed point representation.